File: NeetCode 150/Backtracking/46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        subset = []
        used = [False for _ in range(len(nums))] # tracks which elements are already used

        def backtrack():
            if len(nums) == len(subset):
                res.append(subset[:])
                return
            # Try every number as the next choice
            for i in range(len(nums)):
                if used[i]: # skip if number already used
                    continue
                used[i] = True
                subset.append(nums[i])
                backtrack()
                subset.pop()
                used[i] = False
        
        backtrack()
        return res

File: NeedCode 150 (JAVA)/Backtracking/46. Permutations.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> subset = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        boolean[] used = new boolean[nums.length];
        backtrack(nums,used);
        return res;
    }

    private void backtrack(int[] nums, boolean[] used){
        if(subset.size() == nums.length){
            res.add(new ArrayList<>(subset));
        }

        for(int i =0;i<nums.length;i++){
            if(used[i]) continue;

            used[i] = true;
            subset.add(nums[i]);

            backtrack(nums,used);

            subset.remove(subset.size() - 1);
            used[i] = false;
        }
    }
}
